### PR TITLE
SNOW-619443: Use Column objects for choosing Metadata Columns

### PR DESCRIPTION
**Pull request by sfc-gh-aalam**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes #SNOW-619443

2. Fill out the following pre-review checklist:

   - [x] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

   Use objects from `Column` class to choose metadata columns objects.

### CODE REVIEW

**Review by sfc-gh-jdu — COMMENTED** [review]
Should we add test cases for select?

File: src/snowflake/snowpark/dataframe_reader.py [hunk 1]
```diff
@@ -288,7 +288,7 @@ def schema(self, schema: StructType) -> "DataFrameReader":
         return self
 
     def with_metadata(
-        self, *metadata_cols: Iterable[_MetadataColumn]
+        self, *metadata_cols: Iterable[ColumnOrName]
```
Comment (sfc-gh-yixie) [inline src/snowflake/snowpark/dataframe_reader.py:291]: Do you want to use [typing.Literal](https://docs.python.org/3/library/typing.html#typing.Literal) to limit the input types to those constants?

**Review by sfc-gh-sfan — APPROVED** [review]

File: src/snowflake/snowpark/dataframe_reader.py (hunk 1 above)
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/dataframe_reader.py:291]: Literal only allows `None`, literal values - `int`, `str`, `bytes`, `bool`, or enum values. I would have to create enums for Metadatacolumn which would be more work than necessary. I am happy with `ColumnOrName`.

Comment (sfc-gh-yixie) [reply src/snowflake/snowpark/dataframe_reader.py:291]: OK, forget it then.

**Review by sfc-gh-yixie — APPROVED** [review]

**Pull request merged**